posthog/temporal/data_imports/pipelines/pipeline_non_dlt.py

```diff
@@ -0,0 +1,392 @@
+
+    def to_hogql_types(self) -> dict[str, str]:
+        return self.schema
+
+
+class DeltaTableHelper:
+    _resource_name: str
+    _job: ExternalDataJob
+
+    def __init__(self, resource_name: str, job: ExternalDataJob) -> None:
+        self._resource_name = resource_name
+        self._job = job
+
+    def _get_credentials(self):
+        if TEST:
+            return {
+                "aws_access_key_id": settings.AIRBYTE_BUCKET_KEY,
+                "aws_secret_access_key": settings.AIRBYTE_BUCKET_SECRET,
+                "endpoint_url": settings.OBJECT_STORAGE_ENDPOINT,
+                "region_name": settings.AIRBYTE_BUCKET_REGION,
+                "AWS_ALLOW_HTTP": "true",
+                "AWS_S3_ALLOW_UNSAFE_RENAME": "true",
+            }
+
+        return {
+            "aws_access_key_id": settings.AIRBYTE_BUCKET_KEY,
+            "aws_secret_access_key": settings.AIRBYTE_BUCKET_SECRET,
+            "region_name": settings.AIRBYTE_BUCKET_REGION,
+            "AWS_DEFAULT_REGION": settings.AIRBYTE_BUCKET_REGION,
+            "AWS_S3_ALLOW_UNSAFE_RENAME": "true",
+        }
+
+    def _get_delta_table_uri(self) -> str:
+        normalized_resource_name = NamingConvention().normalize_identifier(self._resource_name)
+        return f"{settings.BUCKET_URL}/{self._job.folder_path()}/{normalized_resource_name}"
+
+    def _evolve_delta_schema(self, schema: pa.Schema) -> deltalake.DeltaTable:
+        delta_table = self.get_delta_table()
+        if delta_table is None:
+            raise Exception("Deltalake table not found")
+
+        delta_table_schema = delta_table.schema().to_pyarrow()
+
+        new_fields = [
+            deltalake.Field.from_pyarrow(field)
+            for field in ensure_delta_compatible_arrow_schema(schema)
+            if field.name not in delta_table_schema.names
+        ]
+        if new_fields:
+            delta_table.alter.add_columns(new_fields)
+
+        return delta_table
+
+    @lru_cache(maxsize=1, condition=lambda result: result is not None)
+    def get_delta_table(self) -> deltalake.DeltaTable | None:
+        delta_uri = self._get_delta_table_uri()
+        storage_options = self._get_credentials()
+
+        if deltalake.DeltaTable.is_deltatable(table_uri=delta_uri, storage_options=storage_options):
+            return deltalake.DeltaTable(table_uri=delta_uri, storage_options=storage_options)
+
+        return None
+
+    def write_to_deltalake(
+        self, data: pa.Table, is_incremental: bool, chunk_index: int, primary_keys: Sequence[Any] | None
+    ) -> deltalake.DeltaTable:
+        delta_table = self.get_delta_table()
+
+        if delta_table:
+            delta_table = self._evolve_delta_schema(data.schema)
+
+        if is_incremental and delta_table is not None:
+            if not primary_keys or len(primary_keys) == 0:
+                raise Exception("Primary key required for incremental syncs")
+
+            delta_table.merge(
```

Comment:
Will function work on an empty delta table? Asking because it'd clean up this logic a bunch if we could just handle `if delta_table is None` before this entire if block

Reply:
Would it work, unsure, likely. But I'd like to keep this separate, delta-rs seems to have a bunch of side effects with these different funcs (and different `modes` etc) - also we should only be using the `merge` function on incremental syncs, things like primary keys are required for incremental syncs but are not for full refresh